### ISSUE TITLE
请问该论文模型支持流式推理？

### ISSUE THREAD

**Issue by hulucky1102** [ioyy900205/MFNet#2]

**Comment by ioyy900205** [comment]
您好，这个模型不支持流式推理。

**Comment by hulucky1102** [comment]
抱歉，再次打扰您，有两个问题想再请假您一下，首先是关于特征使用STDCT,对于最后同样使用mask,STDCT特征对于DFT是否具有更好的性能，其次是PixelShuffleBlock，若该操作仅在特征维度上进行变换，是否能取得比反卷积更优的性能.若您能根据您的经验给予知道，不甚感激。

**Comment by ioyy900205** [comment]
您好，对于第一个问题，我觉得直接从特征的显著性考虑更好。
**特征上**：如果同样做512的STFT和DCT，结果是：257个频带的STFT和512个频带的DCT，所以从分辨率上看，DCT更好。但是STFT包含实部虚部，DCT只有正负（实际上正负就是DCT相位，可以做实验证明），所以从相位上看，DCT似乎是做了一个聚类，显然STFT更好。所以，我感觉是DCT更偏向于突出幅度特征，而把相位特征量化了。
**使用mask**：对于基于mask的方式，实际上是引入了一个inductive bias让网络更容易学习。显然，网络计算量小的情况下，有很强的inductive bias会更好。而DCT也有这种归纳偏置。
所以，我认为 在一般测试环境下，网络计算量有限的情况下，DCT会更好。其他我就不知道了。

第二个问题，假设一个feature是BCTF，如果仅在特征维度（F）上变换，那么可以实现一个因果的版本。你说的反卷积应该是希望pixelshuffle和转置卷积去进行对比吧。这里我不好下结论，仅贴图。来自微软的DeepVQE , 他这里的sub pix conv 就是pixel shuffle操作
![image](https://github.com/ioyy900205/MFNet/assets/76168691/b2f1aabe-5d31-4db1-b17d-806c2aa2e41b)

**Comment by hulucky1102** [comment]
好的好的，感谢大佬指导。

**Comment by hulucky1102** [comment]
能否再请教一个问题，关于CRN模型中的归一化层，若在模型使用BN进行训练，因果模型在实际的使用过程中，音频的长度远远长于训练长度，而且在batchsize较小时，bn权重的var会一直变大。而使用LN归一化层，仅对单帧数据进行归一化操作，这样便导致了前后帧间能量变化被忽略了。是否有一种操作既能保证前后帧间的关系又与音频长度无关。

**Comment by ioyy900205** [comment]
1. 其实训练5秒，测试用2分钟数据不影响 所谓的训练和验证时长不一致的问题。我们可以把网络看作是一个滑窗滤波器，只不过窗是5秒。
2. batch_size较小的时候，确实会造成BN中统计的均值方差 的 var偏大。但是推理用的均值方差使用的是滑动平均之后的均值方差，这个是比较稳定的。Layernorm确实会存在你说的这个问题。我们可以这么考虑，如果feature是BCTF，实际上BN是保留C，在BTF上求均值和方差，而我们共同讨论的Layernorm是在reshape到BT(CF)后，对CF求均值方差，这样T间的能量变化被忽略掉。同样BN中会对C间的能量差忽略，不过这没啥影响。简单来说就是保留哪个维度，就对哪个维度间的差异性进行忽略。
3. 论文中我使用的layernorm是自己手写的。简单来说就是保留了BTF，只在C上reduce。
4. 我自己实验结果显示batch_size小的时候反而收敛会更快一点，确实和理论上batch_size大可以获得更好的结果不一致。不过我也发现了deepfilternet作者的结论跟我一致。可以看到他也是从小batch_size开始训练起来的。
![image](https://github.com/ioyy900205/MFNet/assets/76168691/2755a141-1e14-4ea3-a0a5-d057c5082632)

**Comment by hulucky1102** [comment]
1. 您好，对于您提出的LN存在一些困惑，如上图表述，在通道上做ln操作，是对上图中的[A, a, a_]求均值和方差么？我从tf,keras,layers.LayerNormalization(axis=[-1,-2])更跟到tf,keras,layers.LayerNormalization(axis=[-1])时，模型收敛变得很差。
![image](https://github.com/ioyy900205/MFNet/assets/56753631/3fa86ada-ac7d-405a-b5f2-a69aec749670)
2. 我尝试了DCT作为输入特征，同样的结构下（模型较小），DCT输入特征收敛不如FFT-Mag特征输入，这是否可理解为当模型较小的时候，较少的目标能使模型的收敛变得更加容易。我看论文Rethinking complex-valued deep neural networks for monaural speech enhancement在低参数情况下也是如此。

**Comment by ioyy900205** [comment]
问题1：请参考这段代码哈 layernorm
```
def forward(x, weight, bias, eps):
        N, C, H, W = x.size()
        mu = x.mean(1, keepdim=True)
        var = (x - mu).pow(2).mean(1, keepdim=True)
        y = (x - mu) / (var + eps).sqrt()
        y = weight.view(1, C, 1, 1) * y + bias.view(1, C, 1, 1)
        return y
```
reduce 是在c维度上。同时仿射变换也需要加上。这就是因为你之前提到的，C之间的差异被忽略掉了，所以需要加一个仿射变换给补回来。weight,bias是可以学习的参数，类似BN

问题2：你可以贴一下你这边的DCT变换方式，我看看有没有不同的地方。另外，可能结果还与mask方式噪声强度有关系。同意您的看法，模型较小就需要更强的归纳偏置，也就是您说的较少的目标。像transfomer这种结构归纳偏置就比卷积差，所以需要更大的模型更多的数据。Rethinking complex-valued deep neural networks for monaural speech enhancement 这篇论文我没有找到支持的证据，能否贴出？

**Comment by hulucky1102** [comment]
非常感谢您的回复，问题二在于我表述的不清楚，我理解的时该论文只能支持，当模型较小时，较少的目标有利于模型的训练。

> DCT变换代码如下，窗函数做过对比实验，没有较多的差异性。
![image](https://github.com/ioyy900205/MFNet/assets/56753631/b7d11a4d-94b7-4fe9-b2ff-10ca01c36889)

**Comment by ioyy900205** [comment]
额，这边我是用的是torch做的，我使用的是https://github.com/zh217/torch-dct/tree/master/torch_dct，我看了一下，都是type2类型，应该没有问题。所以，我目前也不知道为什么你这边DCT性能没有stft幅值好的原因了,抱歉。。。ps：记得dctloss中 一定要有abs(dct)，这部分对应着幅值loss

**Comment by hulucky1102** [comment]
非常感谢您耐心的指导
